### DISCUSSION TITLE
Memoization of referencable components

### DISCUSSION THREAD

**Discussion by jpedryc** [Nova-Edge/laravel-openapi#29] [Ideas]
### TLDR
I've investigated initially some memoization possibilities while building components:  
:notebook: https://github.com/TartanLeGrand/laravel-openapi/compare/main...jpedryc:laravel-openapi:caching-tmp-v2

#### User Benefit
Improved performance by removing unnecessary `build()` calls.

## Full story
In our Schema components we're loading a bit of stuff for the descriptions, enums etc. (e.g. from the DB) and it seems for every referencable object all these things are loaded (`build()`) every time. Because the OpenAPI Endpoint is being loaded by an external UI tool and we want it to be working fast and "quiet" (limit external calls while generating the doc), we have limited the required calls by using a simple memoization variable, e.g.

```PHP
class ExampleSchema extends SchemaFactory implements Reusable
{
    public function build(): SchemaContract
    {
        static $buildSchema;

        return $buildSchema ??= Schema::object('ExampleSchema')
            ->properties(
                Schema::integer('some_integer')
                    ->example(6000)
                    ->nullable(),
                /**
                * More schema properties - some bigger, some smaller. Some are loading stuff 
                * from the DB for example.
                */
                SomeBiggerSchema::ref('bigger_sub_schema')
                    ->nullable(),
            );
    }
}
```

#### Started this discussion on two grounds:
1. If someone wants to limit the `build()` behaviour - it has to be done like that in every Schema component :see_no_evil: 
2. **Actually - why shouldn't more people benefit from that** :sweat_smile: 

### What's happening now?

Assuming we're having something like:
```text
FirstExampleResponse: ResponseFactory
  * ExampleSchema::ref('e1'),
  * ExampleSchema::ref('e2'),
  * ExampleSchema::ref('e3'),
SecondExampleResponse: ResponseFactory
  * ExampleSchema::ref('e1'),
  * ExampleSchema::ref('e2'),
  * ExampleSchema::ref('e3'),
```

For this case the `ExampleSchema` component will be build 12 times. In my initial PoC (https://github.com/TartanLeGrand/laravel-openapi/compare/main...jpedryc:laravel-openapi:caching-tmp-v2) it will only once.

What do you think @TartanLeGrand? Would it be of interest and should I investigate/prepare a more reliable PR? :thinking:

**Comment by TartanLeGrand** [comment]
Thank you for your detailed analysis and proposed solution! 🙌 I completely agree with your approach to improving performance by introducing memoization. It aligns perfectly with the goals of making the package faster ⚡ and more developer-friendly 👩‍💻👨‍💻.

I believe there is room for significant optimization, and centralizing this behavior would indeed benefit all Schema components without requiring developers to implement it manually. 🛠️ Additionally, this could open the door for broader discussions on how to streamline the package further.

🚀 Please proceed with preparing a more reliable PR. I’d be happy to collaborate and discuss any additional tasks or improvements needed to ensure we deliver the best version of this package. Looking forward to seeing your proposed changes! 💡

Big thanks 🙏

**Reply by jpedryc** [reply]
Awesome! Will get on making a more reliable PR :wink: :ok_hand:

**Reply by jpedryc** [reply]
:information_source: PR https://github.com/TartanLeGrand/laravel-openapi/pull/30 provided.
